src/main/java/com/jeovan/gymcrmsystem/services/TrainerServiceImpl.java

```diff
@@ -3,49 +3,80 @@
 import com.jeovan.gymcrmsystem.daos.TrainerDao;
 import com.jeovan.gymcrmsystem.models.Trainee;
 import com.jeovan.gymcrmsystem.models.Trainer;
-import com.jeovan.gymcrmsystem.storage.InMemoryStorage;
+import com.jeovan.gymcrmsystem.models.User;
 import org.springframework.beans.factory.annotation.Autowired;
+import org.springframework.security.access.annotation.Secured;
+import org.springframework.security.crypto.password.PasswordEncoder;
 import org.springframework.stereotype.Service;
 
 import java.util.List;
-import java.util.Map;
 import java.util.UUID;
 
 @Service
 public class TrainerServiceImpl implements TrainerService{
-    @Autowired
-    private InMemoryStorage inMemoryStorage;
-
     @Autowired
     private TrainerDao trainerDao;
+    @Autowired
+    private CredentialGeneratorServiceImpl credentialGeneratorService;
+    @Autowired
+    private PasswordEncoder passwordEncoder;
 
     @Override
-    public void setTrainerService() {
-        trainerDao.setStorage((Map<UUID, Trainer>) inMemoryStorage.getStorage().get("Trainer"));
-    }
-
-    @Override
-    public Map<UUID, Trainer> getAll() {
-        return trainerDao.getAll(Trainer.class.getSimpleName());
+    @Secured("ADMIN")
+    public List<Trainer> getAll() {
+        return trainerDao.findAll();
     }
 
     @Override
     public Trainer create(Trainer trainer) {
+        User user = trainer.getUser();
+        user.setUsername(credentialGeneratorService.generateUsername(user.getFirstName(), user.getLastName()));
+        user.setPassword(passwordEncoder.encode(credentialGeneratorService.generatePassword()));
         return trainerDao.save(trainer);
     }
 
     @Override
+    @Secured("ADMIN")
     public Trainer update(Trainer trainer) {
-        return trainerDao.update(trainer);
+        return trainerDao.save(trainer);
     }
 
     @Override
+    @Secured("ADMIN")
     public Trainer select(UUID id) {
-        return trainerDao.getById(id).get();
+        return trainerDao.findById(id).get();
     }
 
+    @Secured("ADMIN")
+    public Trainer selectByUsername(String username) {
+        return trainerDao.findByUsername(username).get();
+    }
+
+    @Override
+    public void deleteByUsername(String username) {
+        trainerDao.deleteByUsername(username);
+    }
+
+    @Secured("ADMIN")
+    public Trainer updatePassword(String username){
```

Comment:
Just a quick question, why updatePassword is not part of the TrainerService interface?, if we want to use this in the future can be convenient have this in the abstraction.

Reply:
I went backwards, first I implemented the method then added to the interface. Fixed in next commit